src/osd/PrimaryLogPG.cc

```diff
@@ -2597,18 +2597,36 @@ int PrimaryLogPG::do_manifest_flush(OpRequestRef op, ObjectContextRef obc, Flush
       }();
       bufferlist in;
       if (fp_oid != tgt_soid.oid) {
-	// decrement old chunk's reference count
-	ObjectOperation dec_op;
-	cls_chunk_refcount_put_op put_call;
-	put_call.source = soid;
-	::encode(put_call, in);
-	dec_op.call("cas", "chunk_put", in);
-	// we don't care dec_op's completion. scrub for dedup will fix this.
-	tid = osd->objecter->mutate(
-	  tgt_soid.oid, oloc, dec_op, snapc,
-	  ceph::real_clock::from_ceph_timespec(obc->obs.oi.mtime),
-	  flags, NULL);
-	in.clear();
+	// decrement old chunk's reference count 
+	SnapSetContext* ssc = obc->ssc;
+	bool has_reference = false;
+	if (ssc && ssc->snapset.clones.size() > 0) {
+	  vector<snapid_t>::reverse_iterator p = ssc->snapset.clones.rbegin();
+	  while (p != ssc->snapset.clones.rend() && *p >= soid.snap)
+	    ++p;
+	  if (p != ssc->snapset.clones.rend()) {
+	    hobject_t next = soid;
+	    next.snap = *p;
+	    ceph_assert(next.snap < soid.snap);
+	    if (check_has_ref_snap(soid, *p, iter->second)) {
+	      has_reference = true;
+	    }
+	  }
+	} 
+	if (!has_reference) {
```

Comment:
Is it actually possible for flush to decrement a reference with the new approach where we don't retain references to "dirty" extents?  It seems like flush should only create new references, right?

Reply:
Yes. But, this PR focus on chunk_info in snapshotted manifest object and refcounting. I'd prefer to make another PR that address "without dirty" implementation.

Reply:
Well, ok.  In that case, I want you to remove this code and leave a TODO indicating that you plan on making this case impossible.

Reply:
Do I need to delete this one now? This is necessary to test refcount in this PR (only based on current implementation, not new approach). How about leaving comment only?

Reply:
It sounds to me like this patch would work differently and be simpler without dirty extents.  I think it would be better to do that first and the push a simpler more correct version of this PR than to add unnecessary complexity that will only be removed later.

Reply:
If so, I want to remove a part of decrementing a reference, and re-push this. Because there are  parts that are not related to reference.

Reply:
I'm happy to take a look.  We can certainly merge a smaller portion of this PR, I simply don't want to merge relatively complex code that we know we won't need.

Reply:
What do you think

Reply:
Did you repush it?

Reply:
Not yet. I'm working on it.

Reply:
Ah, sorry, I misunderstood.  I'll check back in the morning.